lib/src/install/config.rs

```diff
@@ -14,32 +14,120 @@ pub(crate) struct InstallConfigurationToplevel {
+}
+
 /// The serialized [install] section
 #[derive(Debug, Clone, Serialize, Deserialize, Default)]
 #[serde(rename = "install", rename_all = "kebab-case", deny_unknown_fields)]
 pub(crate) struct InstallConfiguration {
     /// Root filesystem type
     pub(crate) root_fs_type: Option<super::baseline::Filesystem>,
+    pub(crate) filesystem: Option<BasicFilesystems>,
     /// Kernel arguments, applied at installation time
     #[serde(skip_serializing_if = "Option::is_none")]
     pub(crate) kargs: Option<Vec<String>>,
 }
 
-impl InstallConfiguration {
-    /// Apply any values in other, overriding any existing values in `self`.
-    fn merge(&mut self, other: Self) {
-        fn mergeopt<T>(s: &mut Option<T>, o: Option<T>) {
-            if let Some(o) = o {
-                *s = Some(o);
+fn merge_basic<T>(s: &mut Option<T>, o: Option<T>) {
+    if let Some(o) = o {
+        *s = Some(o);
+    }
+}
+
+trait Mergeable {
+    fn merge(&mut self, other: Self)
+    where
+        Self: Sized;
+}
+
+impl<T> Mergeable for Option<T>
+where
+    T: Mergeable,
+{
+    fn merge(&mut self, other: Self)
+    where
+        Self: Sized,
+    {
+        if let Some(other) = other {
+            if let Some(s) = self.as_mut() {
+                s.merge(other)
+            } else {
+                *self = Some(other);
             }
         }
-        mergeopt(&mut self.root_fs_type, other.root_fs_type);
+    }
+}
+
+impl Mergeable for RootFS {
+    /// Apply any values in other, overriding any existing values in `self`.
+    fn merge(&mut self, other: Self) {
+        merge_basic(&mut self.fstype, other.fstype)
+    }
+}
+
+impl Mergeable for BasicFilesystems {
+    /// Apply any values in other, overriding any existing values in `self`.
+    fn merge(&mut self, other: Self) {
+        self.root.merge(other.root)
+    }
+}
+
+impl Mergeable for InstallConfiguration {
+    /// Apply any values in other, overriding any existing values in `self`.
+    fn merge(&mut self, other: Self) {
+        merge_basic(&mut self.root_fs_type, other.root_fs_type);
+        self.filesystem.merge(other.filesystem);
         if let Some(other_kargs) = other.kargs {
             self.kargs
                 .get_or_insert_with(Default::default)
                 .extend(other_kargs)
         }
     }
+}
+
+impl InstallConfiguration {
+    /// Some fields can be specified multiple ways.  This synchronizes the values of the fields
+    /// to ensure they're the same.
+    ///
+    /// - install.root-fs-type is synchronized with install.filesystems.root.type; if
+    ///   both are set, then the latter takes precedence
+    pub(crate) fn canonicalize(&mut self) {
+        // New canonical form wins.
+        if let Some(rootfs_type) = self
+            .filesystem
+            .as_ref()
+            .and_then(|r| r.root.as_ref())
+            .and_then(|f| f.fstype.as_ref())
+        {
+            self.root_fs_type = Some(rootfs_type.clone())
```

Comment:
clippy nit: the `Filesystem` enum is `Copy`, so drop the `clone()` and just make a copy with `Some(*rootfs_type)`.  Or maybe better drop `as_ref()` on `fstype` above, then it becomes just `Some(rootfs_type)`.